tests/test_study.py

```diff
@@ -559,3 +559,58 @@ def test_storage_property():
 
     study = optuna.create_study()
     assert study.storage == study._storage
+
+
+@pytest.mark.parametrize('n_jobs', [1, 4])
+def test_callbacks(n_jobs):
+    # type: (int) -> None
+
+    study = optuna.create_study()
+
+    def objective(trial):
+        # type: (optuna.trial.Trial) -> float
+
+        return trial.suggest_int('x', 1, 1)
+
+    # Empty callback list.
+    study.optimize(objective, callbacks=[], n_trials=10, n_jobs=n_jobs)
+
+    # A callback.
+    values = []
+    callbacks = [lambda study, trial: values.append(trial.value)]
+    study.optimize(objective, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert values == [1] * 10
+
+    # Two callbacks.
+    values = []
+    params = []
+    callbacks = [
+        lambda study, trial: values.append(trial.value),
+        lambda study, trial: params.append(trial.params)
+    ]
+    study.optimize(objective, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert values == [1] * 10
+    assert params == [{'x': 1}] * 10
+
+    # If a trial is failed with an exception and the exception is caught by the study,
+    # callbacks are invoked.
+    states = []
+    callbacks = [lambda study, trial: states.append(trial.state)]
+    study.optimize(lambda t: 1/0, callbacks=callbacks, n_trials=10, n_jobs=n_jobs)
+    assert states == [optuna.structs.TrialState.FAIL] * 10
+
+    # NOTE: Because `Study.optimize` blocks forever if `n_jobs` is more than `1` and
```

Comment:
I think we should leave a pointer to the original problem. I raised an issue; so, would you mind mentioning it in the comment?

https://github.com/pfnet/optuna/issues/538

Reply:
Thanks. I agree with you.
Fixed: https://github.com/pfnet/optuna/pull/480/commits/aa930bd95d9f61c2ab20966fdb4a9925b0698fd0